# Steps to achieve the task:

## Step 1: download a clean Linux source tree
```bash
wget https://cdn.kernel.org/pub/linux/kernel/v6.x/linux-6.10.tar.xz
```

## Step 2: Copy the files into the linux/ folder using the command below:
```bash
mkdir linux
tar -xvf linux-6.10.tar.xz -C linux --strip-components=1
```

## Step 3: Copy the linux/ directory into copied_linux/ using rsync
```bash
rsync -auv linux/ copied_linux/
```

## Step 4: I found and truncated all .c files to 0 bytes using the command below:
```bash
find copied_linux/ -name "*.c" -exec truncate -s 0 {} \;
```

## Step 5: Once the .c files are emptied, I generated a patch against the original source to make sure that the entire codebase has been included: 
```bash
diff -ruN linux/ copied_linux/ > all_diff_patch.diff
```

#### <B> Note </b>: The patch generated by the command above may not work correctly because it includes parent directory names in the file paths inside .diff file. 
<i> For example, instead of `arch/alpha/boot/bootp.c`, it results in `linux/arch/alpha/boot/bootp.c`. </i>

<pre>
diff -ruN <span style="color: #d17373;">linux/arch/alpha/boot/bootp.c</span> <span style="color: #d17373;">linux/arch/alpha/boot/bootp.c</span>
--- <span style="color: #d17373;">linux/arch/alpha/boot/bootp.c</span> 2024-09-26 21:55:19.784242880 -0400
+++ <span style="color: #d17373;">linux/arch/alpha/boot/bootp.c</span> 2024-09-28 19:30:23.700222624 -0400
</pre>


#### <b> Problem </b>: Incorrect Directory Paths in Patch File

<i> Because the `koverage` tool requires linux source directory name as input and then concatenates the source file name with the updated files, the resulting paths to be checked by the tools for validity will become even more problematic, such as:
`linux/linux/arch/alpha/boot/bootp.c`. This creates non-existent file paths, causing errors in this part of the code (`koverage.py`):
</i>
```python
91            fullpath = os.path.join(linux_ksrc, srcfile)
92            if not os.path.isfile(fullpath):
93                logger.error("Source file \"%s\" could not be found in the Linux source \"%s\".\n" % (srcfile, linux_ksrc))
94                exit(KOVERAGE_EXITCODE_SOURCEFILE_NOT_FOUND_IN_LINUX_KSRC)
```

#### <b> Solution: </b> Remove Parent Directory Names from Paths
<i>To fix this issue, remove the parent directory names by using `sed` while generating the patch:</i>

```bash
diff -ruN linux/ copied_linux/ | sed 's|^diff -ruN linux/|diff -ruN |; s| copied_linux/| |; s|^--- linux/|--- |; s|^+++ copied_linux/|+++ |' > all_diff_patch.diff
```

## Step 6: Input the file "all_diff_patch.diff" to the koverage tool
```bash
koverage -f --linux-ksrc /home/eshgin/koverage_task/linux --config /home/eshgin/koverage_task/linux/.config --arch x86_64 --check-patch all_diff_patch.diff -o coverage_results_all_task.json 2>&1 | tee mylogs.logs
```

### <b> Problem: </b> Some files caused different kinds of errors when the program ran. 
#### <b> Note: </b> By default, in the error returned, the files causing these errors were not mentioned, so I had to inject print statements into the `koverage.py` file print file names before calling the `get_conditional_blocks` method in the code part below:
```python
620        print(f"The source file is (koverage): {srcfile}")
621        with open(full_srcfpath, 'r') as f:
622            content = f.read()
623        # Read the line count
624        # Old method was to count the '\n' in source content but this failed
625        # for a file (commit 88f8575bca5f, drivers/gpu/drm/amd/amdgpu/gfx_v9_4_2.c)
626        # probably because there was no newline at the end of the file. Use
627        # the file system to count the lines as below, which is safer.
628        with open(full_srcfpath, 'r') as f:
629            line_count = len(f.readlines())
630        
631        # Get the conditional blocks (start-end lines).
632        cb = SyntaxAnalysis.get_conditional_blocks(content, line_count)
```

These files are:
1) The first faulty file was: `arch/powerpc/platforms/powermac/udbg_adb.c`.
    The error it caused:
    ```
    The source file is (koverage): arch/powerpc/platforms/powermac/udbg_adb.c
    Traceback (most recent call last):
      File "/home/eshgin/.local/bin/koverage", line 1120, in <module>
        main()
      File "/home/eshgin/.local/bin/koverage", line 1056, in main
        validation_covreq, validation_patch = get_validation_covreq(covreq, linux_ksrc)
      File "/home/eshgin/.local/bin/koverage", line 637, in get_validation_covreq
        cb = SyntaxAnalysis.get_conditional_blocks(content, line_count)
      File "/home/eshgin/.local/pipx/venvs/kmax/lib/python3.8/site-packages/kmax/superc.py", line 739, in get_conditional_blocks
        assert len(stack) == 1
    AssertionError
    ```
2) The second faulty file was: `drivers/media/pci/cx23885/netup-eeprom.c`.
    The error it caused:
    ```
    The source file is (koverage): drivers/media/pci/cx23885/netup-eeprom.c
    Traceback (most recent call last):
      File "/home/eshgin/.local/bin/koverage", line 1120, in <module>
        main()
      File "/home/eshgin/.local/bin/koverage", line 1056, in main
        validation_covreq, validation_patch = get_validation_covreq(covreq, linux_ksrc)
      File "/home/eshgin/.local/bin/koverage", line 637, in get_validation_covreq
        cb = SyntaxAnalysis.get_conditional_blocks(content, line_count)
      File "/home/eshgin/.local/pipx/venvs/kmax/lib/python3.8/site-packages/kmax/superc.py", line 702, in get_conditional_blocks
        next_token_to_type = categorized_tokens[linenum][i + 1]
    ```
3) The third faulty file was: `drivers/net/can/mscan/mscan.c` 
    ```
    The source file is (koverage): drivers/net/can/mscan/mscan.c
    Traceback (most recent call last):
      File "/home/eshgin/.local/bin/koverage", line 1120, in <module>
        main()
      File "/home/eshgin/.local/bin/koverage", line 1056, in main
        validation_covreq, validation_patch = get_validation_covreq(covreq, linux_ksrc)
      File "/home/eshgin/.local/bin/koverage", line 637, in get_validation_covreq
        cb = SyntaxAnalysis.get_conditional_blocks(content, line_count)
      File "/home/eshgin/.local/pipx/venvs/kmax/lib/python3.8/site-packages/kmax/superc.py", line 702, in get_conditional_blocks
        next_token_to_type = categorized_tokens[linenum][i + 1]
    IndexError: list index out of range
    ```
4) The fourth faulty file was: `drivers/net/ethernet/i825xx/sun3_82586.c`. The error it caused:
    ```
    The source file is (koverage): drivers/net/ethernet/i825xx/sun3_82586.
    Traceback (most recent call last):
      File "/home/eshgin/.local/bin/koverage", line 1120, in <module>
        main()
      File "/home/eshgin/.local/bin/koverage", line 1056, in main
        validation_covreq, validation_patch = get_validation_covreq(covreq, linux_ksrc)
      File "/home/eshgin/.local/bin/koverage", line 637, in get_validation_covreq
        cb = SyntaxAnalysis.get_conditional_blocks(content, line_count)
      File "/home/eshgin/.local/pipx/venvs/kmax/lib/python3.8/site-packages/kmax/superc.py", line 715, in get_conditional_blocks
        stack[-1].sub_block_groups.append([new_cb])
    IndexError: list index out of range
    ```
5) The fifth faulty file was: `init/main.c` The error it caused:
    ```
    The source file is (koverage): init/main.c
    Traceback (most recent call last):
      File "/home/eshgin/.local/bin/koverage", line 1120, in <module>
        main()
      File "/home/eshgin/.local/bin/koverage", line 1056, in main
        validation_covreq, validation_patch = get_validation_covreq(covreq, linux_ksrc)
      File "/home/eshgin/.local/bin/koverage", line 637, in get_validation_covreq
        cb = SyntaxAnalysis.get_conditional_blocks(content, line_count)
      File "/home/eshgin/.local/pipx/venvs/kmax/lib/python3.8/site-packages/kmax/superc.py", line 739, in get_conditional_blocks
        assert len(stack) == 1
    AssertionError
    ```
6) The sixty faulty file was: `drivers/gpu/drm/omapdrm/omap_dmm_tiler.c` 
    The error it caused:
    ```
    The source file is (koverage): drivers/gpu/drm/omapdrm/omap_dmm_tiler.c
    Traceback (most recent call last):
      File "/home/eshgin/.local/bin/koverage", line 1115, in <module>
        main()
      File "/home/eshgin/.local/bin/koverage", line 1051, in main
        validation_covreq, validation_patch = get_validation_covreq(covreq, linux_ksrc)
                                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/eshgin/.local/bin/koverage", line 632, in get_validation_covreq
        cb = SyntaxAnalysis.get_conditional_blocks(content, line_count)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/eshgin/.local/share/pipx/venvs/kmax/lib/python3.12/site-packages/kmax/superc.py", line 737, in get_conditional_blocks
        assert len(stack) == 1
    ```

#### <b> Solution: </b> The minimal solution I could offer to solve these issues was to detect the files causing errors and skip truncating it for the next phase of the process. 
#### <b> Note: </b> Because all the files were experiencing issues in the method, to make faulty file detection faster by avoiding repetition, when I find the faulty file, the next time I just considered the files after the specific faulty file. To do this, I wrote a bash script that takes a specific line in the specified .diff file and remove all the previous lines before that:
```bash
FILE="$1"
LINE_NUMBER="$2"

# Validation stuff...

# Extract lines starting from the next line after the specified line number and overwrite the file
tail -n +"$((LINE_NUMBER + 1))" "$FILE" > temp_file && mv temp_file "$FILE"
```
#### 
1) arch/alpha/kernel/ptrace.c caused error afterwards:
    stderr: "b"patch: **** Can't rename file arch/alpha/kernel/ptrace.c.ohBGgHS to arch/alpha/kernel/ptrace.c : Too many open files\n"". in folder: error2
    


When I omitted the lines before init/main.c, the output was a json file, but when I included all the lines, there were not happening any errors but output file was not present




diff -ruN drivers/net/ethernet/i825xx/sni_82596.c


koverage -f --linux-ksrc /home/eshgin/Desktop/koverage_task/linux --config /home/eshgin/Desktop/koverage_task/linux/.config --arch x86_64 --check-patch fw_all_diff_patch.diff -o /home/eshgin/Desktop/koverage_task/coverage_results_all_task_working.json 2>&1 | tee all_working_logs.log


